### ISSUE TITLE
Feature Request: Twitter Thread Archiver

### ISSUE THREAD

**Issue by shimizurei** [pirate/ArchiveBox#345]
Can something like the Thread Reader App be incorporated into ArchiveBox?

## Type 

 - [x] Propose a brand new feature

## What is the problem that your feature request solves
We can save Twitter threads (NOT individual Twitter posts) as functionally complete articles.

## Describe the ideal specific solution you'd want, and whether it fits into any broader scope of changes
A nice article pdf like the Thread Reader app.

## What hacks or alternative solutions have you tried to solve the problem?
[ThreadReader App](https://threadreaderapp.com/help)

## How badly do you want this new feature?

 - [ ] It's an urgent deal-breaker, I can't live without it
 - [ ] It's important to add it in the near-mid term future
 - [x] It would be nice to have eventually

---

 - [ ] I'm willing to contribute dev time / money to fix this issue
 - [x] I like ArchiveBox so far / would recommend it to a friend
 - [ ] I've had a lot of difficulty getting ArchiveBox set up

**Comment by pirate** [comment]
Yeah I've wanted this for a long time too. The way it's been implemented on other projects is as a content script that unrolls threads before snapshotting inside of chrome headless.

**Comment by mAAdhaTTah** [comment]
Would it be possible for the archiver to trigger the ThreadReader app to unroll it then archive the ThreadReader result?

**Comment by shimizurei** [comment]
Then it ends up depending on ThreadReader. What if ThreadReader becomes defunct tomorrow?

**Comment by mAAdhaTTah** [comment]
@shimizurei You'd have an archive of the ThreadReader page in your ArchiveBox.

**Comment by shimizurei** [comment]
If it's part of ArchiveBox's code, then it's life depends on the maintainers of ArchiveBox. ThreadReader isn't open source, so if it goes down tomorrow, that's it. Everyone will be scrambling to find a replacement because the code is not easily available. Yes, you'll have your already created archives, but you wouldn't be able to create anymore.

**Comment by pirate** [comment]
I'd rather do this via puppeteer scripts once our async playwright worker system is released in v0.6.0 or v0.7.0.

Follow here for updates on puppeteer script support progress: https://github.com/ArchiveBox/ArchiveBox/issues/51

**Comment by akmadian** [comment]
I would really like this feature, and I'm willing to contribute code to make it happen, if that's welcome.

**Comment by pirate** [comment]
There are still a lot of structural blockers in Archivebox's design to running content scripts directly during archiving. 

The most helpful approach might be to write a dedicated extractor in Python that dumps the unrolled thread to a nicer HTML file? Look for existing tools structured like YouTube-dl but for Reddit and Twitter (does a `thread-dl` exist?), and then clone the YOUTUBEDL extractor code to get started.

**Comment by jpaulickcz** [comment]
I've been looking for a box with this functionality for a long while now, with no luck. The closest thing to what I found is https://github.com/weskerfoot/TweetLog however that does require access to developer API which I don't have.

Regular thread - sequence of tweets making a mini article - can be otherwise quite easily archived with threadreaderapp.com (by calling threadreaderapp.com/thread/$URL.html and then downloading it a few minutes later, but I am looking for something that would be able to archive a whole conversations with replies.

**Comment by jpaulickcz** [comment]
I've been looking for a box with this functionality for a long while now, with no luck. The closest thing to what I imagine and that I found is https://github.com/weskerfoot/TweetLog – however that does require access to developer API which I don't have.

Regular thread – sequence of tweets making a mini article (my god, what happened to good ol' blogs?) – can be otherwise quite easily archived with [Thread Reader App](https://threadreaderapp.com/) (by calling `https://threadreaderapp.com/thread/$TWIDENT.html` where `$TWIDENT` is ID of any of the tweet thats part of the thread; and then downloading it a few minutes later. Although I am looking for something that would be able to archive a tweet OR a thread, including all of the replies to one or more of the tweets included in said thread.

**Comment by onemenzel** [comment]
Since threaderapp is now out of service, it would be nice to be able to archive twitter threads as simple html. Maybe there could be a config option to supply a twitter developer token. Then a library like tweepy could be used to download the thread and just put every tweet into a paragraph

**Comment by onemenzel** [comment]
ThreadReaderApp has been acquired by twitter and shut down. I think a feasible approach would be to make a config option where a twitter developer token can be entered and then just download the thread and put it into a simple html file with one ˋ<p>ˋaragraph tag per tweet, maybe ˋ<br> for newlines.

**Comment by pirate** [comment]
How about Nitter?

**Comment by jnolan7599** [comment]
RSS-Bridge, _"the RSS feed for websites missing it",_ advertises the ability to [pull Twitter feeds](https://rss-bridge.github.io/rss-bridge/General/Screenshots.html). Self-hosting, docker & public instances exist. Why couldn't that be [piped into ArchiveBox](https://github.com/ArchiveBox/ArchiveBox/wiki/Usage#Import-a-list-of-URLs-from-a-text-file)?

**Comment by pirate** [comment]
FYI we use Mercury (recently renamed `postlight`) as an extractor already, and they're rapidly adding extractors on their side for many different kinds of sites, so we should get these improvements with no effort required on the archivebox side:

- Reddit threads: https://github.com/postlight/parser/pull/746
- HN threads: https://github.com/postlight/parser/pull/745
- Twitter threads: https://github.com/postlight/parser/pull/622
